### ISSUE TITLE
HP DL380 IPMI Power Meter

### ISSUE THREAD

**Issue by chron0** [collectd/collectd#819]
Ahoy, I've got an issue similar to #735.  I would like to gather all the power meter readings to have metrics showing the power usage of different systems but the ipmi plugin only logs the temperature sensors and doesn't even seem to recognize the power meter at all:

```
Initialization complete, entering read-loop.
ipmi plugin: sensor_list_add: Ignore sensor VRM 2 processor_module (9.2), because I don't know how to handle its type (0x9, power_unit). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor VRM 1 processor_module (9.1), because I don't know how to handle its type (0x9, power_unit). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Power Supplies power_supply (10.3), because I don't know how to handle its type (0x8, power_supply). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Power Supply 2 power_supply (10.2), because I don't know how to handle its type (0x8, power_supply). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Power Supply 1 power_supply (10.1), because I don't know how to handle its type (0x8, power_supply). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Ext. Health LED system_chassis (23.3), because I don't know how to handle its type (0xc0, invalid). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Int. Health LED system_chassis (23.2), because I don't know how to handle its type (0xc0, invalid). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor UID Light system_chassis (23.1), because I don't know how to handle its type (0xc0, invalid). If you need this sensor, please file a bug report.
EINF: mydomain(0.20)(s,0) sdr.c(handle_sdr_data): SDR fetch error getting sdr 0x0: c1

# ipmitool sdr
UID Light        | 0 unspecified     | ok
Int. Health LED  | 0 unspecified     | ok
Ext. Health LED  | 0 unspecified     | ok
Power Supply 1   | 0 unspecified     | nc
Power Supply 2   | 0 unspecified     | nc
Power Supplies   | 0 unspecified     | nc
VRM 1            | 0 unspecified     | cr
VRM 2            | 0 unspecified     | cr
Fan 1            | 58.02 unspecifi | nc
Fan 2            | 58.02 unspecifi | nc
Fan 3            | 50.57 unspecifi | nc
Fan 4            | 53.70 unspecifi | nc
Fan 5            | 53.70 unspecifi | nc
Fan 6            | 53.70 unspecifi | nc
Fans             | 0 unspecified     | nc
Temp 1           | 47 degrees C      | ok
Temp 2           | 28 degrees C      | ok
Temp 3           | 55 degrees C      | ok
Temp 4           | 55 degrees C      | ok
Temp 5           | 57 degrees C      | ok
Temp 6           | 55 degrees C      | ok
Temp 7           | 55 degrees C      | ok
Power Meter      | 430 Watts         | cr


# ipmitool sensor
UID Light        | 0.000      | unspecified | ok    | na        | na        | na        | na        | na        | na        
Int. Health LED  | 0.000      | unspecified | ok    | na        | na        | na        | na        | na        | na        
Ext. Health LED  | 0.000      | unspecified | ok    | na        | na        | na        | na        | na        | na        
Power Supply 1   | 0.000      | unspecified | nc    | na        | na        | na        | na        | na        | na        
Power Supply 2   | 0.000      | unspecified | nc    | na        | na        | na        | na        | na        | na        
Power Supplies   | 0.000      | unspecified | nc    | na        | na        | na        | na        | na        | na        
VRM 1            | 0.000      | unspecified | cr    | na        | na        | na        | na        | na        | na        
VRM 2            | 0.000      | unspecified | cr    | na        | na        | na        | na        | na        | na        
Fan 1            | 58.016     | unspecified | nc    | na        | na        | na        | na        | na        | na        
Fan 2            | 58.016     | unspecified | nc    | na        | na        | na        | na        | na        | na        
Fan 3            | 50.568     | unspecified | nc    | na        | na        | na        | na        | na        | na        
Fan 4            | 53.704     | unspecified | nc    | na        | na        | na        | na        | na        | na        
Fan 5            | 53.704     | unspecified | nc    | na        | na        | na        | na        | na        | na        
Fan 6            | 53.704     | unspecified | nc    | na        | na        | na        | na        | na        | na        
Fans             | 0.000      | unspecified | nc    | na        | na        | na        | na        | na        | na        
Temp 1           | 47.000     | degrees C  | ok    | 0.000     | 0.000     | 0.000     | 67.000    | 70.000    | 75.000    
Temp 2           | 28.000     | degrees C  | ok    | 0.000     | 0.000     | 0.000     | 37.000    | 39.000    | 44.000    
Temp 3           | 55.000     | degrees C  | ok    | 0.000     | 0.000     | 0.000     | 68.000    | 127.000   | 127.000   
Temp 4           | 55.000     | degrees C  | ok    | 0.000     | 0.000     | 0.000     | 68.000    | 127.000   | 127.000   
Temp 5           | 57.000     | degrees C  | ok    | 0.000     | 0.000     | 0.000     | 74.000    | 77.000    | 82.000    
Temp 6           | 55.000     | degrees C  | ok    | 0.000     | 0.000     | 0.000     | 68.000    | 127.000   | 127.000   
Temp 7           | 55.000     | degrees C  | ok    | 0.000     | 0.000     | 0.000     | 68.000    | 127.000   | 127.000   
Power Meter      | 424.000    | Watts      | cr    | na        | na        | na        | na        | na        | na        
```

In the meantime, is there a quick hack to get the data, other than collecting the value from ipmitool sdr via shell script? If you need more data, just tell me. Thanks a lot. BTW, octo, we still have a lunch date to fulfill.

**Comment by rrajeevan** [comment]
Hi @chron0 , out of curiosity, were you able to get a hack working?

**Comment by chron0** [comment]
Hihihi, no, I didn't. and in the meantime I've migrated to a G7 and retired the G5, so I didn't have that problem anymore. Sorry.

**Comment by rrajeevan** [comment]
Thanks for the reply, @chron0 . So, are you able to monitor Power Supply using collectd in G7?

**Comment by chron0** [comment]
I'd say yes, these are from right now, on a DL-380 G7. I guess this was just a problem with the old G5 hardware and there are probably not so many G5s in production anymore. 

```
# ipmitool sdr
UID Light        | 0x00              | ok
Sys. Health LED  | 0x00              | ok
Power Supply 1   | 35 Watts          | ok
Power Supply 2   | 155 Watts         | ok
Power Supplies   | 0x00              | ok
Fan 1            | 43.51 percent     | ok
Fan 2            | 43.51 percent     | ok
Fan 3            | 41.55 percent     | ok
Fan 4            | 49 percent        | ok
Fan 5            | 49 percent        | ok
Fan 6            | 13.72 percent     | ok
Fans             | 0x00              | ok
Temp 1           | 16 degrees C      | ok
Temp 2           | 40 degrees C      | ok
Temp 3           | 40 degrees C      | ok
Temp 4           | 27 degrees C      | ok
Temp 5           | 27 degrees C      | ok
Temp 6           | 30 degrees C      | ok
Temp 7           | 34 degrees C      | ok
Temp 8           | 35 degrees C      | ok
Temp 9           | 31 degrees C      | ok
Temp 10          | 38 degrees C      | ok
Temp 11          | 29 degrees C      | ok
Temp 12          | 35 degrees C      | ok
Temp 13          | 28 degrees C      | ok
Temp 14          | 29 degrees C      | ok
Temp 15          | 28 degrees C      | ok
Temp 16          | 23 degrees C      | ok
Temp 17          | 26 degrees C      | ok
Temp 18          | 26 degrees C      | ok
Temp 19          | 22 degrees C      | ok
Temp 20          | 24 degrees C      | ok
Temp 21          | 28 degrees C      | ok
Temp 22          | 28 degrees C      | ok
Temp 23          | 34 degrees C      | ok
Temp 24          | 29 degrees C      | ok
Temp 25          | 27 degrees C      | ok
Temp 26          | 28 degrees C      | ok
Temp 27          | 27 degrees C      | ok
Temp 28          | 27 degrees C      | ok
Temp 29          | 35 degrees C      | ok
Temp 30          | 62 degrees C      | ok
Memory           | 0x00              | ok
Power Meter      | 184 Watts         | ok
```

```
# ipmitool sensor
UID Light        | 0x0        | discrete   | 0x0080| na        | na        | na        | na        | na        | na        
Sys. Health LED  | 0x0        | discrete   | 0x0080| na        | na        | na        | na        | na        | na        
Power Supply 1   | 35         | Watts      | ok    | na        | na        | na        | na        | na        | na        
Power Supply 2   | 165        | Watts      | ok    | na        | na        | na        | na        | na        | na        
Power Supplies   | 0x0        | discrete   | 0x0180| na        | na        | na        | na        | na        | na        
Fan 1            | 43.120     | percent    | ok    | na        | na        | na        | na        | na        | na        
Fan 2            | 43.120     | percent    | ok    | na        | na        | na        | na        | na        | na        
Fan 3            | 41.552     | percent    | ok    | na        | na        | na        | na        | na        | na        
Fan 4            | 49         | percent    | ok    | na        | na        | na        | na        | na        | na        
Fan 5            | 49         | percent    | ok    | na        | na        | na        | na        | na        | na        
Fan 6            | 13.720     | percent    | ok    | na        | na        | na        | na        | na        | na        
Fans             | 0x0        | discrete   | 0x0180| na        | na        | na        | na        | na        | na        
Temp 1           | 16.000     | degrees C  | ok    | na        | na        | na        | na        | 41.000    | 45.000    
Temp 2           | 40.000     | degrees C  | ok    | na        | na        | na        | na        | 82.000    | 83.000    
Temp 3           | 40.000     | degrees C  | ok    | na        | na        | na        | na        | 82.000    | 83.000    
Temp 4           | 27.000     | degrees C  | ok    | na        | na        | na        | na        | 87.000    | 92.000    
Temp 5           | 27.000     | degrees C  | ok    | na        | na        | na        | na        | 87.000    | 92.000    
Temp 6           | 29.000     | degrees C  | ok    | na        | na        | na        | na        | 87.000    | 92.000    
Temp 7           | 33.000     | degrees C  | ok    | na        | na        | na        | na        | 87.000    | 92.000    
Temp 8           | 35.000     | degrees C  | ok    | na        | na        | na        | na        | 90.000    | 95.000    
Temp 9           | 31.000     | degrees C  | ok    | na        | na        | na        | na        | 65.000    | 70.000    
Temp 10          | 38.000     | degrees C  | ok    | na        | na        | na        | na        | 90.000    | 95.000    
Temp 11          | 29.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 12          | 35.000     | degrees C  | ok    | na        | na        | na        | na        | 90.000    | 95.000    
Temp 13          | 28.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 14          | 29.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 15          | 28.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 16          | 23.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 17          | 26.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 18          | 26.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 19          | 22.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 20          | 24.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 21          | 28.000     | degrees C  | ok    | na        | na        | na        | na        | 80.000    | 85.000    
Temp 22          | 28.000     | degrees C  | ok    | na        | na        | na        | na        | 80.000    | 85.000    
Temp 23          | 34.000     | degrees C  | ok    | na        | na        | na        | na        | 77.000    | 82.000    
Temp 24          | 29.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 25          | 27.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 26          | 28.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 27          | 27.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 28          | 27.000     | degrees C  | ok    | na        | na        | na        | na        | 70.000    | 75.000    
Temp 29          | 35.000     | degrees C  | ok    | na        | na        | na        | na        | 60.000    | 65.000    
Temp 30          | 62.000     | degrees C  | ok    | na        | na        | na        | na        | 110.000   | 115.000   
Memory           | 0x0        | discrete   | 0x4080| na        | na        | na        | na        | na        | na        
Power Meter      | 184        | Watts      | ok    | na        | na        | na        | na        | na        | na  
```

**Comment by rrajeevan** [comment]
Thanks for the reply, @chron0 . So is collectd picking up the following values?
`Power Supply 1   | 35         | Watts      | ok    | na        | na        | na        | na        | na        | na `  
`Power Supply 2   | 165        | Watts      | ok    | na        | na        | na        | na        | na        | na `


If yes, which version of collectd are you using? I tried collectd5.5 on a G8 but it still gives me the "**Ignore sensor Power Supply  power_supply**" error. 
I patched src/ipmi.c to the following which gets rid of the error but power is still not getting reported :( 

```
case IPMI_SENSOR_TYPE_POWER_SUPPLY:
  type = "power";
  break;

**Comment by chron0** [comment]
Ah, sorry, my bad, context switch was a bit too abrupt and I was only thinking about ipmitool and not how collectd deals with it. A quick manual collectd run returned:

```
Initialization complete, entering read-loop.
ipmi plugin: sensor_list_add: Ignore sensor Memory system_board (7.8), because I don't know how to handle its type (0xc, memory). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Power Supplies power_supply (10.3), because I don't know how to handle its type (0x8, power_supply). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Power Supply 2 power_supply (10.2), because I don't know how to handle its type (0x8, power_supply). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Power Supply 1 power_supply (10.1), because I don't know how to handle its type (0x8, power_supply). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor Sys. Health LED system_chassis (23.2), because I don't know how to handle its type (0xc0, invalid). If you need this sensor, please file a bug report.
ipmi plugin: sensor_list_add: Ignore sensor UID Light system_chassis (23.1), because I don't know how to handle its type (0xc0, invalid). If you need this sensor, please file a bug report.
EINF: mydomain(0.20)(s,0) sdr.c(handle_sdr_data): SDR fetch error getting sdr 0x0: c1
```
So, the prior yes turns into a: No, it doesn't work on the G7 either. Tested withcollectd-5.4.1-r2 and 5.5.1. Sorry for the confusion but this issue has been raised so long ago...

**Comment by rrajeevan** [comment]
Thanks for clarifying! Any luck with a workaround? 
I tried to use exec plugin but that won't work as collectd doesn't like executing as root. sudo su the only option?

http://mailman.verplant.org/pipermail/collectd/2014-March/006117.html

**Comment by chron0** [comment]
Seems like it :/ - Alternatively you could let a cron script run to periodically fetch the data as root and put the data output into a temp file which is then read by collectd. I'm doing the same with conky for metrics a regular user cannot grab (like nvidia-smi data) but in both cases I find that to be the least elegant option.

**Comment by varesa** [comment]
I have a similar issue with a G6. I worked around it using the exec plugin with sudo as suggested above. Another option would have been an udev rule to set wider permissions on /dev/ipmi0.

Here is the script (python) I used to parse the output of "ipmitool sdr": https://gist.github.com/varesa/2c473b70ffd771e6ffa071654c970aaa

**Comment by rpv-tomsk** [comment]
New features was added to trunk and will be released in collectd-5.8, please check if they are good for you. Feel free to (re)open tickets if needed.
